aspnetcore/performance/memory.md

```diff
@@ -0,0 +1,408 @@
+    fp.Watch("*.*");
+}
+```
+
+[PhysicaFileProvider](/dotnet/api/microsoft.extensions.fileproviders.physicalfileprovider?view=dotnet-plat-ext-3.0) is a managed class, so any instance will be collected at the end of the request.
+
+Here is the resulting memory profile while invoking this API continuously.
+
+![preceding chart](memory/_static/fileprovider.png)
+
+The preceding chart shows an obvious issue with the implementation of this class, as it keeps increasing memory usage. This is a known problem that is being tracked in [this issue](https://github.com/aspnet/Home/issues/3110).
+
+The same leak could be happen in user code, by one of the following:
+
+* Not releasing the class correctly.
+* Forgetting to invoke the `Dispose`method of the dependent objects that should be disposed.
+
+#### Large objects heap
+
+Frequent memory allocation/free cycles can fragment memory, especially when allocating large chunks of memory. Objects are allocated in contiguous blocks of memory. To mitigate fragmentation, when the GC frees memory, it will try to defragment it. This process is called **compaction**. Compaction involves moving objects. Moving large objects imposes a performance penalty. For this reason the GC creates a special memory zone for  _large_ objects, called the [large object heap](/dotnet/standard/garbage-collection/large-object-heap) (LOH). Object that are greater than 85,000 bytes (approximately 83 KB) are:
+
+* Placed on the LOH.
+* Not compacted.
+* Eventually released during generation 2 collections.
+
+When the LOH is full, the GC will trigger an automatic generation 2 collection. Generation 2 collections:
+
+* Are inherently slow.
+* Additionally incur the cost of triggering a collection on all other generations.
+
+The following API that illustrates this behavior:
+
+```csharp
+[HttpGet("loh/{size=85000}")]
+public int GetLOH1(int size)
+{
+    return new byte[size].Length;
+}
+```
+
+The following chart shows the memory profile of calling the `/api/loh/84975` endpoint, under maximum load:
+
+![preceding chart](memory/_static/loh1.png)
+
+The following chart shows the memory profile of calling the `/api/loh/84976` endpoint, allocating *just one more byte*:
+
+![preceding chart](memory/_static/loh2.png)
+
+Note: The `byte[]` structure has overhead bytes. That's why 84,976 bytes triggers the 85,000 limit.
+
+Comparing the two preceding charts:
+
+* The working set is similar for both scenarios, about 450 MB.
+* The under LOH requests (84,975 bytes) shows mostly generation 0 collections.
+* The over LOH requests generate constant generation 2 collections. Generation 2 collections are expensive. More CPU is required and throughput drops almost 50%.
+
+For maximum performance, large object use should be minimized. If possible, split up large objects. For example, [Response Caching](xref:performance/caching/response) middleware in ASP.NET Core split the cache entries into blocks less than  85,000 bytes.
```

Comment:
>For maximum performance, large object use should be minimized. If possible, split up large objects. For  [](start = 0, length = 104)

should emphasize that temporary large objects are a problem - they are the ones that cause gen2 GCs. see "Collection cost" in [https://devblogs.microsoft.com/dotnet/large-object-heap-uncovered-from-an-old-msdn-article/](https://devblogs.microsoft.com/dotnet/large-object-heap-uncovered-from-an-old-msdn-article/)